### ISSUE TITLE
[Feat. Req.] Pixiv website support

### ISSUE THREAD

**Issue by momoe** [Simon-Tesla/RaccoonyWebEx#26]
As the title suggests, getting http://www.pixiv.net support would be another major artist's outlet that Raccooney could support. Handles very much the same way Hentai-Foundry works. (And wonderful job on HF by  the way!)

The only issue is some of the galleries submissions are nested like InkBunny and some submissions are animated using some weird javascript with a sequence of PNGs. 

Please consider adding pixiv support to RaccooneyWebEx. 

-TY~

**Comment by momoe** [comment]
Thanks for taking the time and energy to work at this. I know it's one of the big ones. 

P.S. do you have a  paypal to donate to? I'd love to show my support. You've been doing the lord's work.

**Comment by Eupeptic1** [comment]
I am poking around at this.

**Huge disclaimer:** I can't read, write, understand, or speak Japanese.  I can figure some of it out by context (the big bold text is the title, the smaller and longer text is the description); some of the page elements have English names in the HTML; I know where Google Translate is.  But I am inevitably going to guess wrong about something.

Doing a "view source" on a complete submission page shows that there's a giant meta tag in the `<head>` - it looks like it has *all* of the information on the submission as JSON.  The `<body>` section is curiously blank, when viewed this way; I think maybe a script is reading that big meta tag and rendering the page?  Not sure.

When selecting *parts* of the complete submission page, and doing a "view source", things like the title, description seem to be in divs or headings (h1, h2, etc) with consistent class names.  The class names aren't English words (example: `class="sc-khSPhy idKHda"`), but they seem to be consistent across submissions.

Early design decision: the tags all seem to be listed in kanji (猫), and many of them seem to have romaji (neko) and English (cat) translations.  Which one do we want to save in Raccoony?  Or do we want to save all the versions of each tag, so that both English and Japanese speakers can get something they can understand?

**Comment by Eupeptic1** [comment]
I have a very alpha version working.  It can find the metadata and the links to the images for a single submission.  Downloading doesn't work, and it doesn't yet know what to do with a gallery/favorites/index page.  I have also only tested it against still images, and not stories, animations, etc.

Points of note:

In the JSON provided by Pixiv, the key for the data on the current image is the submission ID.  Right now, I parse that ID out of the current page's URL, and then proceed to use it as a key to the JSON array.  I feel like I should do some kind of sanitize or trim or other cleanup before I do that, but I don't know exactly what.

In other words, for a URL of https://www.pixiv dot net/en/artworks/12345 , the JSON looks like
```javascript
{
  "illust": {
    "12345": {
      "illustId": "12345",
      "illustTitle": "My nifty artwork",
      "illustComment": "This is a cool picture I drew",
      "id": "12345"
    }
  }
}
```
and I then do things like
```javascript
let urlParts = pageUrl.split("/");
let submissionId = urlParts.pop();
let title = postData["illust"][submissionId]["illustTitle"];
```
to get the various pieces of metadata.

They seem to be doing some kind of referrer checking, or otherwise limiting downloads if you aren't logged in.  Without even using Raccoony, I can load a page, pick out the thumbnail URL from the page source, paste that URL in a new tab, and it will load.  However, if I hit "reload" in that tab with the thumbnail, I immediately get a 403.  When using Raccoony, it gets far enough that the raccoon head offers the "Download" button, but if you click it, the download never goes through.  The fix may be to specify that you have to be logged in to your Pixiv account for Raccoony to work.

More in the next few days.

**Comment by Simon-Tesla** [comment]
DeviantArt support has a similar problem with its download links. I haven't been able to figure out a satisfactory solution, so I end up downloading the images that are on the page, which generally works (presumably because FF is able to yank it out of the browser cache instead of re-downloading it).

**Comment by Eupeptic1** [comment]
The download problem seems to be different than what I thought.  I don't think it's the referrer checking.

The way a Raccoony plugin seems to work, is that when you first visit a submission page, it looks at the page and pulls out all the data it wants.  If it finds everything, the main part of Raccoony offers the Download button.

If you then click the Download button, the plugin then *parses the whole page again*.  If it finds everything, it can pass the URL of the image to the main part of Raccoony, and the main part kicks off the download.  If it doesn't find everything on this second pass, the download won't work.

I *think* that Pixiv might only send the big meta tag the first time you load an image.  It's not there on the second try, which makes the download fail in Raccoony.  (Or, possibly, some script at Pixiv puts the meta tag in the page, and I'm not always waiting long enough for that to happen.)

The flaw in this theory: I added a `logger.log(document);` to the plugin, and I can't see the big meta tag in *either* case, even when I visit images I've never visited before ever.  If my theory was correct, I would expect to see the big meta tag in the first case, and not see it in the second case.

I know the plugin thinks the big meta tag is there, at least the first time, because it wouldn't be able to parse things from it, if it wasn't there.

I am still working on it.

**Comment by Eupeptic1** [comment]
Whatever Pixiv is doing is smarter than I am.

Fun test: don't even have Raccoony installed.  In Firefox, visit an image on Pixiv, such as https://www.pixiv.net/en/artworks/57194842 (sfw).  Right-click and select "view page source".  In the source tab, about halfway down the page on a typical desktop screen (it's right after the "bug bounty" console print), you will see the big meta tag: `<meta name="preload-data" id="meta-preload-data" content="[a bunch of stuff]">`.  It goes on for a couple of screenfuls, at least.

Now, go back to the tab with the image, Tools > Web Developer > Inspector.  Expand the `<head>` section of the tree, and look for that meta tag.

IT ISN'T THERE.

The closest you can get: when viewing source, one of the elements right before the big meta tag is a `<link rel="alternate" type="application/json+oembed" href="https://www.pixiv.net/oembed/?url=[stuff]>`.  If you read down the list in the Inspector, you *can* find that same link (it's right after an `apple-touch-icon` and a `manifest`).  The big meta tag should be right underneath that in the inspector... but it's not.

There *are* a bunch of `<script>` links right after that point in the Inspector.  I actually tried loading all of them manually, to see if any of them populate that big meta tag, and they don't seem to.  (Or at least, they can tell that they're not being loaded by a page on their own site, and change their behavior accordingly.)

The part I am confuse about is that Raccoony can clearly see that big meta tag - *sometimes*.  On the initial page load, it sees it, and can pull out all the metadata.  On the reload that happens when you hit the "Download" button, it can't.

Two things I tried:

- Disabling permission for cookies and local storage for Pixiv, in hopes of getting a "new world every click" kind of response.  No change.

- Putting a loop in the Pixiv plugin: try three times to get the meta tag, sleeping between each attempt, just in case I was racing one of the site scripts.  Sleeping 50 milliseconds between attempts didn't work; sleeping 500 milliseconds between attempts didn't work.  The plugin always got the meta tag the first time, and never got it when I clicked "Download".

There may be a logical reason that I don't understand yet, but I sort of wonder if Pixiv is doing something clever on the server side, specifically to make things like Raccoony *not* work.

So I think I have to give up on the nicely-preformatted metadata, and go back to parsing individual elements out of the HTML, which will break horribly every year or two.  Sigh.  :(

**Comment by Simon-Tesla** [comment]
It looks like they're using the Single Page App model for development where the initial HTML is almost entirely replaced with new DOM after the script loads, which would account for the major discrepancies between the view-source and the inspector. View-source reports what HTML got loaded from the server, not what the current DOM looks like. DeviantArt has a similar model which is why I have a mutationSelector parameter in the plugin constructor--it tells Raccoony to watch for changes that might indicate when it needs to re-parse the page for different data.

It sounds like there's a race condition between when Raccoony starts running its content script on the page and when Pixiv's scripts remove that meta tag from the DOM. In theory we could potentially have some script that runs as soon as possible and caches the JSON from that element so that subsequent queries can use it, but I suspect that the data in that tag will become more stale as the user starts doing things on the site, like clicking to other pictures and such.

**Comment by Eupeptic1** [comment]
Okay, at least I'm not crazy.  :)

Would the Single Page App thing align with the fact that when you do view-source, the entire body of the page is just `<body><div id='root'></div></body>` , and then the various scripts are filling it in later?

I tried to rewrite the plugin to use `querySelector()` to pick out pieces of the HTML, as I see them on the rendered page in the browser.  It didn't find *anything*.  For a test, I did this as the very first thing inside `getMedia()`:

```javascript
        let staticbody = document.body.innerHTML;
        logger.log("pixiv: document body", document.body);
        logger.log("pixiv: document static body", staticbody);
```

The first log statement gives me a populated body in the web console... but I think it's some kind of *reference* that the browser is filling in later, with the current DOM.

All I get from that second log statement in the web console is `<div id="root"></div><div id="raccoonyExtensionRoot"></div>` - that's it, nothing else filled in.

Is it possible to tell the plugin to wait for all of Pixiv's scripts to run, and *then* run the plugin?  Or is that brittle, because they run in a random order, or the last script today might not be the last script tomorrow?

Looking at the DA plugin, I guess the mutationSelector fires every time the HTML body changes, so in effect, the last run of the plugin is after the last script has messed with the body, and hopefully at that point, all the data the plugin wants should be in the page.  It's the same net effect, just that the plugin may run a few times before everything settles.

I also note that the DA plugin uses async functions to actually scrape.  I guess that's related to possibly being repeatedly called when the body changes.

I thought of the caching thing as well, but I wasn't sure how to implement it.  Also, usually if you have a cache, you need a "dump cache" button for when it screws up.

**Comment by Simon-Tesla** [comment]
> Would the Single Page App thing align with the fact that when you do view-source, the entire body of the page is just `<body><div id='root'></div></body>`, and then the various scripts are filling it in later?

Yep. From what I can tell it also takes over browser navigation as well so that when you're clicking around it only pulls metadata from the server and replaces the necessary bits of DOM with the new view, avoiding a page refresh.

Regarding logging, the console generally only logs DOM references and doesn't capture the DOM state at the time of loading; you'd need to query the specific state you're interested in when you log it.

> I also note that the DA plugin uses async functions to actually scrape.

The interface for the function is async, since it returns a `Promise` object; all plugins are implemented with an async API even if they return synchronously (which they do if you don't see an `await` or a `new Promise()` call somewhere). I use the `async` keyword as syntactic sugar to make it so that I can write the function in a more natural synchronous style without having to new up a Promise object myself (`async` does that for me).

> Is it possible to tell the plugin to wait for all of Pixiv's scripts to run, and then run the plugin?

In theory we could wait for the document's `load` event. Right now Raccoony just waits for `DOMContentLoaded` which means the HTML has finished downloading but that referenced script and images and such haven't finished downloading. Not sure if that waits for script to finish its initial run or not though; the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event) only promises that it's finished downloading by the point that it's fired. I wouldn't want to change the behavior for other plugins, so we'd either need to write some support for it in the plugin loader, or we'd need to write the plugin so that it doesn't resolve until after the load event fires and we can parse the page then. (I'd prefer the latter.)

> Looking at the DA plugin, I guess the mutationSelector fires every time the HTML body changes, so in effect, the last run of the plugin is after the last script has messed with the body, and hopefully at that point, all the data the plugin wants should be in the page. It's the same net effect, just that the plugin may run a few times before everything settles.

Yep. If this causes a lot of noise on the pixiv end (which it might, though I already have some debounce logic on the page change handler to avoid the worst of it) I've also just used a setTimeout to delay the parse (see the FurryNetwork plugin for an example of that). Not elegant, but if mutationSelector on its own doesn't work, that might be necessary.

> I thought of the caching thing as well, but I wasn't sure how to implement it. Also, usually if you have a cache, you need a "dump cache" button for when it screws up.

In this case I'm referring to the browser cache; Raccoony doesn't do any real caching of its own. There should be a way to clear it via the dev tools or browser UI.

**Comment by Eupeptic1** [comment]
**tl;dr** I'm working on cloning and hacking the DA plugin and life is better; the Pixiv plugin gets called repeatedly but now it can find the stuff it's looking for.

> Regarding logging, the console generally only logs DOM references and doesn't capture the DOM state at the time of logging; you'd need to query the specific state you're interested in when you log it.

`printf()` is lying to me.  These are dark times indeed. :D

> or we'd need to write the plugin so that it doesn't resolve until after the load event fires and we can parse the page then.

Right now I'm basing it on the DA plugin.  On an initial test, when loading a brand-new page at Pixiv in a new browser session, the Pixiv plugin gets called six times.  It isn't able to find the metadata it wants until the third time.  (It's not looking for all the metadata yet, though.)

I'm not going to implement any logic like "after the third time we get called, it's probably OK to parse"; I was just curious how many times it would get called.  I do have a "give up" fairly early (bail if we can't find the author name), so it doesn't spend a lot of time trying to parse if the page isn't filled in yet.

> > I thought of the caching thing as well, but I wasn't sure how to implement it.

> In this case I'm referring to the browser cache; Raccoony doesn't do any real caching of its own.

What I was thinking of is some kind of internal-to-Raccoony caching.  Like, grab that huge metadata tag when we first see it, and then stick it in some kind of persistent storage (browser local storage?), with a key of the submission ID.  Then, when we get called again, first look for the huge metadata tag in the "live" document, and then if it's not there, go look it up by submission ID in the persistent storage.  I think this would need its own UI for dumping the internal-to-Raccoony cache, and it might also require you to ask for more permissions at plugin install time, which might be a bad idea.

**Comment by Simon-Tesla** [comment]
> What I was thinking of is some kind of internal-to-Raccoony caching

Ehhh... lots of things can go wrong with internal caches. Unless you can think of a foolproof way to invalidate the cache if it's incorrect, I'd rather not go down that route. (The persistent storage part of the problem is trivial; we already have storage that we're currently using for settings, so it doesn't require new permissions.)

**Comment by Eupeptic1** [comment]
Alpha version is working; it can find all the metadata and download images, at least for the three images I tried it on.  I want to dogfood it a bit more before posting a gist.

One interesting thing: you can load a submission page, get the raccoon head, click Download, and the image (and metadata file, if requested) will download; that's all OK.

If you then click on one of the thumbnails on the page (which are other images in that artist's gallery), it behaves as you said earlier - Pixiv rewrites the page to display the new image, without a page reload.  This includes the URL displayed in the browser address bar - it shows the new submission ID.

However, main Raccoony seems to think you're still on the same image; if you click on the raccoon head, the first button is the whited-out "Downloaded", rather than the blue "Download" button.  Looking in the console log, the plugin is running again on the new image, finding the new metadata, and sending that to main Raccoony.

So I think I need some mechanism to tell main Raccoony "hey, the page changed, offer the download button again".  I'm not sure if this mechanism already exists, or needs to be invented.

> Ehhh... lots of things can go wrong with internal caches.

I didn't think it was a *great* idea, but it was something that came to mind before I understood how the site worked.  I never tried to actually implement it.

**Comment by Simon-Tesla** [comment]
Yeah, in theory the mutationSelector parameter should handle cases like this, but it's not super well-tested since most sites Raccoony supports don't do the single-page app thing. Pixiv might require some architecture rework to properly support it.

**Comment by Eupeptic1** [comment]
> So I think I need some mechanism to tell main Raccoony "hey, the page changed, offer the download button again".

I have a version of this working, but I need feedback on whether it's a good idea or not.

I added an element to `AppState` called `lastDownloadId` in definitions.ts.  I also added it to the constructor for the `Page` class in page.tsx; it's initialized to null.

Also in page.tsx, I changed `downloadMedia()` as follows.  First, I added a check on the download ID to the "bail out" statement at the top.  Second, on successful download, I remember the ID from the current media struct.

```javascript
    downloadMedia = (force: boolean = false) => {
        this.props.siteActions.getMedia().then((media) => {
            const { downloadState } = this.state;
            if ( !force &&
                 (downloadState === E.DownloadState.InProgress ||
                  downloadState === E.DownloadState.Done ||
                  downloadState === E.DownloadState.Exists) &&
                 (media.submissionId === this.state.lastDownloadId) ) {
                // FIXME pixiv
                // If downloading has already started, AND we're on the
                // same submission ID, then don't try to re-download.
                return;
            }

            this.setState({ downloadState: E.DownloadState.InProgress });
            sendDownloadMediaMessage(media)
                .then((download: I.DownloadResponse) => {
                    // The download promise resolves when the download starts, not when it finishes.
                    // May want to figure out how to get the download end event.
                    // FIXME pixiv change - remember ID of successful download
                    this.setState({
                        downloadState: download.success ? E.DownloadState.Done : E.DownloadState.Error,
                        lastDownloadId: download.success ? media.submissionId : null,
                    })
                    setTimeout(() => this.setState({
                        downloadState: download.success ? E.DownloadState.Exists : E.DownloadState.NotDownloaded,
                    }), 2000);
                })
        });
    }
```

Finally, in pageOverlay.tsx, I took out the `disabled:true` in the `DownloadState.Exists` state.

What the user sees:

* At the beginning of the world, when you visit a page on Pixiv, the raccoon head offers the "Download" button.

* If you click the "Download" button, the current submission downloads.

* The text of the button changes to "Downloaded", but it does not white out.

* If you then click on a thumbnail of some other submission on the page, Pixiv rewrites the page to show you that other submission.

* The first button offered by the raccoon head says "Downloaded", but if you go ahead and click that button, Raccoony downloads the current submission.

* It goes on like that as you visit more submissions; the download button label doesn't accurately reflect the state of the world anymore, but clicking it causes the current submission to download.

If this is OK, the next action is to figure out how to make the label reflect the state of the world.

I also still have to get the "open gallery in tabs" function working.  Right now, the plugin just works on individual submissions.

**Comment by Simon-Tesla** [comment]
Sorry, I haven't had time to look at this, will try to make some time this weekend. My high-level thoughts is that we should be able to make the plugin send an unambiguous signal that the currently viewed submission has changed, possibly through some combination of ID change and page URL change. The UI code for Raccoony is a bit baroque and due for a refactor, but I think the general approach would be for it to take that change the plugin reports and have it reinitialize with the new page state when it detects that happening.

**Comment by Eupeptic1** [comment]
I will upload what I have to my fork in a minute, so you can grab it and build it if you want.  It still works as I described a couple of days ago; the "download" button keeps working, even after it has changed to "downloaded".

The signaling between the plugin and the main code is something you understand better than I do.

I wonder if part of the puzzle would be to decouple the status reporting from the text of the Download button - in other words, have some separate indicator in the popup menu that tells you what the download is doing.  I am thinking of a spot next to the download button that's initially blank, says "run" and is yellow while a download is in progress, says "ok" and is green when the download completes, and maybe says "trouble" and is red if the download fails.  There might need to be an interlock such that when a download is in progress, the Download button is disabled, but I think it has to be enabled again as soon as we have some disposition on the status of the download.

On the other hand, maybe this is getting too far into duplicating the functionality of the browser's existing "downloading" indicator.  I think it's useful for there to be *some* indication on the popup menu that Raccoony heard you click Download and is doing something about it, but maybe it doesn't need to be red/yellow/green elaborate.

**Comment by Eupeptic1** [comment]
> I will upload what I have to my fork in a minute, 

Uploaded now, on my master.

**Comment by photonometric** [comment]
I'm not sure if this program might contain useful code (python) or bug solutions that may be helpful in developing a WebEx plugin—or perhaps might just be a useful heads-up of a parallel project:

[PixivUtil2](https://github.com/Nandaka/PixivUtil2) is an excellent CLI program for downloading Pixiv images/artists/tags/searches/fanbox subscriptions/etc. It has been in very active development for 8 years and is quite functional and sophisticated. 

Very highly recommended, and thought it worth mentioning here to perhaps avoid reinventing the wheel in some way :3
